blocks/graph.py

```diff
@@ -37,6 +37,7 @@ def _get_variables(self):
         self.applies = set()
         self.application_calls = set()
         self.updates = []
+        self.auxiliary_variables = set()
```

Comment:
Let me start from this point:

> Although I appreciate the argument about splitting things up into parts, I also don't like the idea of 
> some regularizers being done this way, and others in yet another way.

I think the situation here is very similar to the one with models: I believe there is no such thing as "the regularization" and there is nothing in common between different regularization. Examples:
dropout is a blunt force insertion in the computation graph, L2 is addition of a term to the cost, matrix norm constraint is a change in the training algorithm and so on and so forth. Thus we can not have a common interfaces for all regularizations possible. What we should do though, is to provide tools to build new methods and have a set of implementation of most common ones. 

Next, back to weight decay: there was a bit of my fault here as I forgot that biases are typically not included in the weight decay. But that in fact even strengthens my point: do you want every brick to put three auxiliary variables with roles `WEIGHT_DECAY`, `WEIGHT_DECAY_WITH_BIASES` and `BIASES_DECAY` into the computation graph? And code of every brick would have to know about all regularizations supported by the library.

Instead, I find your idea of having additional roles for parameters more attractive. 

So let's go from a simple cases to more complicated ones:
```python
mlp = MLP(activations=[Tanh() for _ in range(4)], dimensions=...)
y_hat = mlp.apply(x)
cost = CategoricalCrossEntropy().apply(y, y_hat)

# Case 1: I am a machine learning newbie. I want weight decay and do not want to know too much 
# about the way your library works.
cost = cost + weight_decay(cost=cost)
# The call above simply selects all parameters tagged as WEIGHTS and returns sum of their square norms.

# Case 2: My weights are OK, but my biases are too large. What shall I do?
cost = cost + weight_decay(cost=cost, roles=[BIASES])

# Case 3: I only care about the third layer of my MLP.
cost = cost + weight_decay(
      params=[param for param in Selector([
                   mlp.linear_transformations[2]]).get_params().values()
              if param.role == WEIGHTS])
# Note: Jan's new selectors should make it look better.
# Alternatively we can provide an interface
cost = cost + weight_decay(bricks=[mlp.linear_transformations[2]])
# which is a short-cut to the whole list comprehension above.
```

Also about difference between `Selector` and `ComputationGraph`: the former deals with the bricks hierarchy and pathes in it, whereas for the latter the world is flat. Computation graph might have routines like `get_variables` from your example, but it should never treat in recursive fashion.
E.g. if I have an MLP and every layer of it adds something to the graph, then in order to exclude variables belonging to all its ancestor I would write

```python
weight_norms = cg.get_variables(roles=[VariableRole.USEFUL_SOMETHING],
                                exclude_bricks=Selector(mlp1).select('//').get_params().values())
```

although this is currently not implemented.